File: backend/app/services/parse_service.py
from __future__ import annotations

from dataclasses import dataclass
from decimal import Decimal
from pathlib import Path

from fastapi import HTTPException, UploadFile, status
from sqlalchemy.orm import Session

from app.config import get_settings
from app.models import ParseJob, User
from app.output.template_writer import output_filename, write_foreign_uplift
from app.parsers.base import ParseError
from app.parsers.registry import get_parser
from app.storage import delete_file, save_upload, stored_original_path, stored_output_path


PDF_MIME = "application/pdf"
XLSX_MIME = "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
ACCEPTED_EXTENSIONS = {".pdf": PDF_MIME, ".xlsx": XLSX_MIME}


@dataclass(frozen=True)
class ParseServiceResult:
    job: ParseJob
    output_filename: str
    output_path: Path


async def parse_upload(
    *,
    db: Session,
    user: User,
    file: UploadFile,
    vendor: str,
    parser_slug: str,
    fx_rate: Decimal,
    margin: Decimal,
) -> ParseServiceResult:
    parser = _resolve_parser(parser_slug, vendor)
    _validate_upload_type(file, parser.accepted_mime)

    source_filename = Path(file.filename or "quote").name
    source_path = stored_original_path(source_filename)
    output_path = stored_output_path()

    try:
        await save_upload(file, source_path, max_bytes=get_settings().max_upload_bytes)
        result = parser.parse(source_path)
        write_foreign_uplift(
            result.line_items,
            output_path,
            margin=margin,
            fx_rate=fx_rate,
            vendor_name=vendor.upper(),
            currency=result.metadata.currency,
        )
    except ValueError as exc:
        delete_file(source_path)
        delete_file(output_path)
        raise HTTPException(status_code=status.HTTP_413_REQUEST_ENTITY_TOO_LARGE, detail="File is too large.") from exc
    except ParseError as exc:
        delete_file(source_path)
        delete_file(output_path)
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail={"stage": exc.stage, "hint": exc.hint, "message": str(exc)},
        ) from exc
    except Exception as exc:
        delete_file(source_path)
        delete_file(output_path)
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail={"stage": "parse", "hint": "Could not parse this file.", "message": str(exc)},
        ) from exc

    job = ParseJob(
        user_id=user.id,
        vendor=vendor,
        parser_slug=parser.slug,
        source_filename=source_filename,
        source_path=str(source_path),
        output_path=str(output_path),
        fx_rate=fx_rate.quantize(Decimal("0.0001")),
        margin=margin.quantize(Decimal("0.01")),
        computed_total=result.validation.computed_total,
        quoted_total=result.validation.quoted_total,
        totals_match=result.validation.matches,
    )
    user.fx_rate = job.fx_rate
    user.margin = job.margin
    db.add(job)
    db.add(user)
    db.commit()
    db.refresh(job)
    return ParseServiceResult(job=job, output_filename=output_filename(source_filename), output_path=output_path)


def _resolve_parser(parser_slug: str, vendor: str):
    try:
        parser = get_parser(parser_slug)
    except KeyError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Unknown parser.") from exc
    if parser.vendor != vendor:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Parser does not match vendor.")
    return parser


def _validate_upload_type(file: UploadFile, expected_mime: str) -> None:
    filename = file.filename or ""
    suffix = Path(filename).suffix.lower()
    by_extension = ACCEPTED_EXTENSIONS.get(suffix)
    if by_extension is None:
        raise HTTPException(status_code=status.HTTP_415_UNSUPPORTED_MEDIA_TYPE, detail="Only PDF and XLSX files are supported.")
    if by_extension != expected_mime:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="File extension does not match selected parser.")

File: src/BidParser.Infrastructure/Services/ParseService.cs
using BidParser.Application.Output;
using BidParser.Application.Parsing;
using BidParser.Domain.Abstractions;
using BidParser.Domain.Models;
using BidParser.Infrastructure.Entities;
using BidParser.Infrastructure.Persistence;
using BidParser.Infrastructure.Storage;
using Microsoft.Extensions.Logging;

namespace BidParser.Infrastructure.Services;

/// <summary>
/// Orchestrates a single parse end to end: resolve + validate the parser and file, persist the upload,
/// run the parser, dispatch to the matching output writer, then record the outcome. Success writes a
/// <see cref="ParseJob"/> + <see cref="ParseMetric"/> (and, on a totals mismatch, a best-effort monitoring
/// entry). A "detect"-stage failure is reclassified as a wrong-file-type selection — nothing is persisted,
/// the upload is deleted, and the likely-correct type is suggested via sibling <c>Detect()</c>; other
/// exceptions are recorded as <see cref="FailedParseJob"/>s. See AGENTS.md "Wrong file-type handling".
/// </summary>
public sealed class ParseService(
    QuoteParseService quoteParseService,
    WorkbookWriteService workbookWriteService,
    FileStorage storage,
    AppDbContext db,
    FailedParseJobRecorder failureRecorder,
    ILogger<ParseService> logger)
{
    /// <summary>
    /// Parses <paramref name="fileStream"/> with the parser identified by <paramref name="parserSlug"/>
    /// and writes the standardised output for <paramref name="crmTemplate"/> (defaults to the parser's).
    /// Returns the output path + validation on success; throws <c>ParseError("fileType")</c> when the
    /// selected parser does not recognise the file (wrong file-type selection).
    /// </summary>
    public async Task<ParseServiceResult> ParseAsync(
        User user,
        Stream fileStream,
        string uploadFilename,
        string vendor,
        string parserSlug,
        decimal? fxRate,
        decimal? margin,
        decimal? imPercent,
        decimal? onCostPercent,
        string? crmTemplate,
        bool includeSubComponentDetail,
        bool splitBySolutionId,
        long maxUploadBytes,
        CancellationToken ct = default)
    {
        QuoteParseSelection selection;
        try
        {
            selection = quoteParseService.ResolveSelection(uploadFilename, vendor, parserSlug);
        }
        catch (ParseInputException inputError)
        {
            throw ToValidationException(inputError);
        }

        var importType = selection.IsAuto ? ImportType.Auto : ImportType.Manual;
        IParser? parser = selection.Parser;

        var displayFilename = Path.GetFileName(uploadFilename);
        var sourcePath = storage.NewOriginalPath(displayFilename);
        var outputPath = storage.NewOutputPath(splitBySolutionId ? ".zip" : ".xlsx");

        await storage.SaveUploadAsync(fileStream, sourcePath, maxUploadBytes, ct);

        // Effective values used by writers and persisted on the ParseJob/ParseMetric ledger.
        // The User's defaults are only updated when the caller explicitly supplied a value
        // (see User mutation block below), so omitting margin from an HP No Calculation parse
        // doesn't clobber a saved value.
        var effectiveFxRate = fxRate ?? 1m;
        var effectiveMargin = margin ?? 0m;

        try
        {
            var parsedQuote = await quoteParseService.ParseAsync(
                new QuoteParseRequest(
                    sourcePath,
                    vendor,
                    parserSlug,
                    new ParseOptions { IncludeSubComponentDetail = includeSubComponentDetail },
                    displayFilename),
                selection,
                ct);
            parser = parsedQuote.Parser;
            var result = parsedQuote.Result;

            var writeResult = await workbookWriteService.WriteAsync(
                new WorkbookWriteRequest(
                    parsedQuote,
                    outputPath,
                    crmTemplate,
                    effectiveFxRate,
                    effectiveMargin,
                    imPercent,
                    onCostPercent,
                    splitBySolutionId),
                ct);
            var template = writeResult.CrmTemplate;
            var outputWorkbookCount = writeResult.WorkbookCount;

            var fxRateRounded = Math.Round(effectiveFxRate, 4, MidpointRounding.AwayFromZero);
            var marginRounded = Math.Round(effectiveMargin, 2, MidpointRounding.AwayFromZero);

            // Bid metadata is best-effort, so a broken anchor degrades silently by design. Log it:
            // this line, and `bid_number IS NULL` on rows created after deployment, are the only
            // signals that a vendor changed a header and a format needs its anchor revisited.
            if (result.Metadata.BidNumber is null)
            {
                logger.LogWarning(
                    "No bid metadata extracted for {Filename} using {Slug} — the history row will show no bid number",
                    displayFilename, parser.Slug);
            }

            var job = new ParseJob
            {
                UserId = user.Id,
                Vendor = vendor,
                ParserSlug = parser.Slug,
                CrmTemplate = template,
                SourceFilename = displayFilename,
                BidNumber = result.Metadata.BidNumber,
                BidRevision = result.Metadata.BidRevision,
                SourcePath = sourcePath,
                OutputPath = outputPath,
                FxRate = fxRateRounded,
                Margin = marginRounded,
                ComputedTotal = result.Validation.ComputedTotal,
                QuotedTotal = result.Validation.QuotedTotal,
                TotalsMatch = result.Validation.Matches,
                SplitBySolutionId = splitBySolutionId,
                ImportType = importType,
            };

            var metric = new ParseMetric
            {
                UserId = user.Id,
                UserUsername = user.Username,
                UserName = user.Name,
                Vendor = vendor,
                ParserSlug = parser.Slug,
                SourceFilename = displayFilename,
                Currency = result.Metadata.Currency,
                QuotedTotal = result.Validation.QuotedTotal,
                ComputedTotal = result.Validation.ComputedTotal,
                TotalsMatch = result.Validation.Matches,
                FxRate = fxRateRounded,
                Margin = marginRounded,
                ImportType = importType,
                ParseJob = job,
            };

            // User defaults: only the last-used vendor is remembered. fx_rate / margin /
            // im_percent are intentionally NOT persisted — the user must enter them
            // every parse so a stale value never gets silently applied.
            user.DefaultVendor = vendor;
            db.Update(user);
            db.Add(job);
            db.Add(metric);
            await db.SaveChangesAsync(ct);

            if (!result.Validation.Matches)
            {
                // Record a monitoring entry so admins can review mismatches.
                // This is best-effort: we swallow any recorder failure so the
                // user's successful parse is unaffected.
                try
                {
                    await failureRecorder.RecordMismatchAsync(
                        user, vendor, parser.Slug, displayFilename, sourcePath,
                        fxRateRounded, marginRounded,
                        result.Validation.ComputedTotal, result.Validation.QuotedTotal,
                        importType,
                        ct);
                }
                catch (Exception ex)
                {
                    // Monitoring record failed — parse job is already committed, continue.
                    logger.LogError(ex,
                        "Failed to record validationMismatch monitoring entry for parse job {ParseJobId} (user {UserId}, parser {ParserSlug})",
                        job.Id, user.Id, parser.Slug);
                }
            }

            var cancelledLines = result.LineItems
                .Where(i => i.IsCancelled)
                .Select(i => new CancelledLine(i.LineSequence ?? string.Empty, i.Vpn))
                .ToList();

            return new ParseServiceResult(
                job,
                writeResult.OutputFilename,
                outputPath,
                outputWorkbookCount,
                result.Validation,
                result.Metadata.Currency,
                cancelledLines,
                result.HasRebateIneligibleItems);
        }
        catch (ParseInputException inputError) when (
            inputError.Kind is ParseInputErrorKind.AutoDetectionFailed or ParseInputErrorKind.WrongFileType)
        {
            // The selected parser could not recognise the file's layout (missing table
            // anchor or required column) — almost always a wrong file-type selection, not
            // a genuine parser failure. Suggest the correct type by comparing the file
            // against the other formats for the SAME vendor, drop everything (no monitoring
            // entry, no parse job/metric) and delete the stored upload.
            storage.TryDelete(outputPath);
            storage.TryDelete(sourcePath);

            logger.LogInformation(
                "Wrong file type for {Filename}: selected {Slug}, suggested {Suggested}",
                displayFilename, parser?.Slug ?? parserSlug, inputError.SuggestedParserName ?? "(none)");

            throw new ParseError("fileType", inputError.Message, inputError.Message);
        }
        catch (ParseInputException inputError) when (inputError.Kind != ParseInputErrorKind.MagicByteMismatch)
        {
            storage.TryDelete(outputPath);
            storage.TryDelete(sourcePath);
            throw ToValidationException(inputError);
        }
        catch (ParseValidationException)
        {
            // User-input 400 (unknown/unsupported CRM template, missing IM%) — a
            // client error, not a genuine parse failure. Drop both files and don't
            // record a FailedParseJob so it doesn't show up as monitoring noise.
            storage.TryDelete(outputPath);
            storage.TryDelete(sourcePath);
            throw;
        }
        catch (Exception ex)
        {
            storage.TryDelete(outputPath);

            var fxRateRounded = Math.Round(effectiveFxRate, 4, MidpointRounding.AwayFromZero);
            var marginRounded = Math.Round(effectiveMargin, 2, MidpointRounding.AwayFromZero);

            var recordedException = ex is ParseInputException { Kind: ParseInputErrorKind.MagicByteMismatch } inputError
                ? new ParseError("upload", inputError.Message, inputError.Message)
                : ex;

            await failureRecorder.RecordAsync(
                user, vendor, parser?.Slug ?? parserSlug, displayFilename, sourcePath,
                fxRateRounded, marginRounded, recordedException, importType, ct);

            if (!ReferenceEquals(recordedException, ex))
            {
                throw recordedException;
            }

            throw;
        }
    }

    private static ParseValidationException ToValidationException(ParseInputException error)
        => new(
            error.Kind == ParseInputErrorKind.UnsupportedExtension ? 415 : 400,
            error.Message);
}

/// <summary>The outcome of a successful parse handed back to the API: the persisted job, the download filename + path, validation, currency, and parse warnings.</summary>
public sealed record ParseServiceResult(
    ParseJob Job,
    string OutputFilename,
    string OutputPath,
    int OutputWorkbookCount,
    ValidationResult Validation,
    string Currency,
    IReadOnlyList<CancelledLine> CancelledLines,
    bool HasRebateIneligibleItems);

/// A line item flagged as cancelled in the source document.
/// Surfaced to the frontend via the <c>X-Cancelled-Lines</c> response header.
/// </summary>
public sealed record CancelledLine(string Line, string Vpn);

/// <summary>A client-input error (bad parser/vendor/template, missing IM%) carrying an HTTP status; not a recorded parse failure.</summary>
public sealed class ParseValidationException : Exception
{
    public ParseValidationException(int statusCode, string detail) : base(detail)
    {
        StatusCode = statusCode;
        Detail = detail;
    }

    public int StatusCode { get; }
    public string Detail { get; }
}
